// react-native-libsodium/lib/module/lib.native.js
export { base64_variants, to_string } from './libsodium-js-utils';
import { base64_variants, to_string } from './libsodium-js-utils';
import { convertToOutputFormat } from './utils';
import { NativeModules } from 'react-native';
const Libsodium = NativeModules.Libsodium;
if (Libsodium && typeof Libsodium.install === 'function') {
  console.log('calling Libsodium.install');
  Libsodium.install();
} else if (!Libsodium) {
  console.warn('Libsodium module not defined');
} else {
  console.warn('Libsodium.install not a function');
}
export const crypto_auth_BYTES = global.jsi_crypto_auth_BYTES;
export const crypto_auth_KEYBYTES = global.jsi_crypto_auth_KEYBYTES;
export const crypto_secretbox_KEYBYTES = global.jsi_crypto_secretbox_KEYBYTES;
export const crypto_secretbox_NONCEBYTES = global.jsi_crypto_secretbox_NONCEBYTES;
export const crypto_box_PUBLICKEYBYTES = global.jsi_crypto_box_PUBLICKEYBYTES;
export const crypto_box_SECRETKEYBYTES = global.jsi_crypto_box_SECRETKEYBYTES;
export const crypto_box_NONCEBYTES = global.jsi_crypto_box_NONCEBYTES;
export const crypto_box_SEEDBYTES = global.jsi_crypto_box_SEEDBYTES;
export const crypto_aead_xchacha20poly1305_ietf_KEYBYTES = global.jsi_crypto_aead_xchacha20poly1305_ietf_KEYBYTES;
export const crypto_aead_xchacha20poly1305_ietf_NPUBBYTES = global.jsi_crypto_aead_xchacha20poly1305_ietf_NPUBBYTES;
export const crypto_kdf_KEYBYTES = global.jsi_crypto_kdf_KEYBYTES;
export const crypto_kdf_CONTEXTBYTES = global.jsi_crypto_kdf_CONTEXTBYTES;
export const crypto_generichash_BYTES = global.jsi_crypto_generichash_BYTES;
export const crypto_generichash_BYTES_MIN = global.jsi_crypto_generichash_BYTES_MIN;
export const crypto_generichash_BYTES_MAX = global.jsi_crypto_generichash_BYTES_MAX;
export const crypto_generichash_KEYBYTES = global.jsi_crypto_generichash_KEYBYTES;
export const crypto_generichash_KEYBYTES_MIN = global.jsi_crypto_generichash_KEYBYTES_MIN;
export const crypto_generichash_KEYBYTES_MAX = global.jsi_crypto_generichash_KEYBYTES_MAX;
export const crypto_sign_SEEDBYTES = global.jsi_crypto_sign_SEEDBYTES;
export const crypto_pwhash_SALTBYTES = global.jsi_crypto_pwhash_SALTBYTES;
export const crypto_pwhash_ALG_DEFAULT = global.jsi_crypto_pwhash_ALG_DEFAULT;
export const crypto_pwhash_OPSLIMIT_INTERACTIVE = global.jsi_crypto_pwhash_OPSLIMIT_INTERACTIVE;
export const crypto_pwhash_MEMLIMIT_INTERACTIVE = global.jsi_crypto_pwhash_MEMLIMIT_INTERACTIVE;
export const crypto_pwhash_BYTES_MIN = global.jsi_crypto_pwhash_BYTES_MIN;
export const crypto_pwhash_BYTES_MAX = global.jsi_crypto_pwhash_BYTES_MAX;
export const _unstable_crypto_kdf_hkdf_sha256_BYTES_MAX = global.jsi_crypto_kdf_hkdf_sha256_BYTES_MAX;
export const _unstable_crypto_kdf_hkdf_sha256_BYTES_MIN = global.jsi_crypto_kdf_hkdf_sha256_BYTES_MIN;
export const _unstable_crypto_kdf_hkdf_sha256_KEYBYTES = global.jsi_crypto_kdf_hkdf_sha256_KEYBYTES;
export const from_base64 = (input, variant) => {
  const variantToUse = variant || base64_variants.URLSAFE_NO_PADDING;
  const result = global.jsi_from_base64_to_arraybuffer(input, variantToUse);
  return new Uint8Array(result);
};
export const to_base64 = (input, variant) => {
  const variantToUse = variant || base64_variants.URLSAFE_NO_PADDING;
  const inputParam = typeof input === 'string' ? input : input.buffer;
  return global.jsi_to_base64(inputParam, variantToUse);
};
export function to_hex(input) {
  const inputParam = typeof input === 'string' ? input : input.buffer;
  return global.jsi_to_hex(inputParam);
}
export function randombytes_buf(length, outputFormat) {
  const result = global.jsi_randombytes_buf(length);
  return convertToOutputFormat(result, outputFormat);
}
export function randombytes_uniform(upper_bound) {
  return global.jsi_randombytes_uniform(upper_bound);
}
export function crypto_auth(message, key, outputFormat) {
  const messageParam = typeof message === 'string' ? message : message.buffer;
  const result = global.jsi_crypto_auth(messageParam, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_auth_keygen(outputFormat) {
  const result = global.jsi_crypto_auth_keygen();
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_auth_verify(tag, message, key) {
  const messageParam = typeof message === 'string' ? message : message.buffer;
  return global.jsi_crypto_auth_verify(tag.buffer, messageParam, key.buffer);
}
export function crypto_secretbox_keygen(outputFormat) {
  const result = global.jsi_crypto_secretbox_keygen();
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_aead_xchacha20poly1305_ietf_keygen(outputFormat) {
  const result = global.jsi_crypto_aead_xchacha20poly1305_ietf_keygen();
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_kdf_keygen(outputFormat) {
  const result = global.jsi_crypto_kdf_keygen();
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_box_keypair(outputFormat) {
  const result = global.jsi_crypto_box_keypair();
  return {
    keyType: 'x25519',
    publicKey: convertToOutputFormat(result.publicKey, outputFormat),
    privateKey: convertToOutputFormat(result.secretKey, outputFormat)
  };
}
export function crypto_box_seed_keypair(seed, outputFormat) {
  const result = global.jsi_crypto_box_seed_keypair(seed.buffer);
  return {
    keyType: 'x25519',
    publicKey: convertToOutputFormat(result.publicKey, outputFormat),
    privateKey: convertToOutputFormat(result.secretKey, outputFormat)
  };
}
export function crypto_sign_keypair(outputFormat) {
  const result = global.jsi_crypto_sign_keypair();
  return {
    keyType: 'ed25519',
    publicKey: convertToOutputFormat(result.publicKey, outputFormat),
    privateKey: convertToOutputFormat(result.secretKey, outputFormat)
  };
}
export function crypto_sign_seed_keypair(seed, outputFormat) {
  const result = global.jsi_crypto_sign_seed_keypair(seed.buffer);
  return {
    keyType: 'ed25519',
    publicKey: convertToOutputFormat(result.publicKey, outputFormat),
    privateKey: convertToOutputFormat(result.secretKey, outputFormat)
  };
}
export function crypto_sign_detached(message, privateKey, outputFormat) {
  let result;
  const messageParam = typeof message === 'string' ? message : message.buffer;
  result = global.jsi_crypto_sign_detached(messageParam, privateKey.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_sign_verify_detached(signature, message, publicKey) {
  let result;
  const messageParam = typeof message === 'string' ? message : message.buffer;
  result = global.jsi_crypto_sign_verify_detached(signature.buffer, messageParam, publicKey.buffer);
  return result;
}
export function crypto_secretbox_easy(message, nonce, key, outputFormat) {
  let result;
  const messageParam = typeof message === 'string' ? message : message.buffer;
  result = global.jsi_crypto_secretbox_easy(messageParam, nonce.buffer, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_secretbox_open_easy(ciphertext, nonce, key, outputFormat) {
  let result;
  const ciphertextParam = typeof ciphertext === 'string' ? ciphertext : ciphertext.buffer;
  result = global.jsi_crypto_secretbox_open_easy(ciphertextParam, nonce.buffer, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_box_easy(message, nonce, publicKey, privateKey, outputFormat) {
  let result;
  const messageParam = typeof message === 'string' ? message : message.buffer;
  result = global.jsi_crypto_box_easy(messageParam, nonce.buffer, publicKey.buffer, privateKey.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_box_open_easy(ciphertext, nonce, publicKey, privateKey, outputFormat) {
  let result;
  const ciphertextParam = typeof ciphertext === 'string' ? ciphertext : ciphertext.buffer;
  result = global.jsi_crypto_box_open_easy(ciphertextParam, nonce.buffer, publicKey.buffer, privateKey.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_box_seal(ciphertext, publicKey, outputFormat) {
  let result;
  const ciphertextParam = typeof ciphertext === 'string' ? ciphertext : ciphertext.buffer;
  result = global.jsi_crypto_box_seal(ciphertextParam, publicKey.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_box_seal_open(ciphertext, publicKey, privateKey, outputFormat) {
  let result;
  const ciphertextParam = typeof ciphertext === 'string' ? ciphertext : ciphertext.buffer;
  result = global.jsi_crypto_box_seal_open(ciphertextParam, publicKey.buffer, privateKey.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_generichash(hash_length, message, key, outputFormat) {
  const messageParam = typeof message === 'string' ? message : message.buffer;
  const result = global.jsi_crypto_generichash(hash_length, messageParam, key ? key.buffer : undefined);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_pwhash(keyLength, password, salt, opsLimit, memLimit, algorithm, outputFormat) {
  if (salt.length !== crypto_pwhash_SALTBYTES) {
    throw new Error('invalid salt length');
  }
  let result;
  const passwordParam = typeof password === 'string' ? password : password.buffer;
  result = global.jsi_crypto_pwhash(keyLength, passwordParam, salt.buffer, opsLimit, memLimit, algorithm);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_kdf_derive_from_key(subkey_len, subkey_id, ctx, key, outputFormat) {
  const result = global.jsi_crypto_kdf_derive_from_key(subkey_len, subkey_id, ctx, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_aead_xchacha20poly1305_ietf_encrypt(message, additional_data, _secret_nonce, public_nonce, key, outputFormat) {
  let result;
  const messageParam = typeof message === 'string' ? message : message.buffer;
  if (typeof additional_data !== 'string') {
    throw new Error('crypto_aead_xchacha20poly1305_ietf_encrypt: input type not yet implemented');
  }
  result = global.jsi_crypto_aead_xchacha20poly1305_ietf_encrypt(messageParam, additional_data, public_nonce.buffer, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function crypto_aead_xchacha20poly1305_ietf_decrypt(_secret_nonce, ciphertext, additional_data, public_nonce, key, outputFormat) {
  let result;
  if (typeof ciphertext === 'string') {
    throw new Error('crypto_aead_xchacha20poly1305_ietf_decrypt: input type not yet implemented');
  }
  if (typeof additional_data !== 'string') {
    throw new Error('crypto_aead_xchacha20poly1305_ietf_decrypt: input type not yet implemented');
  }
  const ciphertextParam = typeof ciphertext === 'string' ? ciphertext : ciphertext.buffer;
  result = global.jsi_crypto_aead_xchacha20poly1305_ietf_decrypt(ciphertextParam, additional_data, public_nonce.buffer, key.buffer);
  return convertToOutputFormat(result, outputFormat);
}
export function _unstable_crypto_kdf_hkdf_sha256_extract(key, salt) {
  return new Uint8Array(global.jsi_crypto_kdf_hkdf_sha256_extract(key.buffer, salt.buffer));
}
export function _unstable_crypto_kdf_hkdf_sha256_expand(key, info, length) {
  return new Uint8Array(global.jsi_crypto_kdf_hkdf_sha256_expand(key.buffer, info, length));
}

// add no-op ready to match the libsodium-wrappers API
export const ready = new Promise(resolve => resolve());

// add no-op ready to match the react-nativ-libsodium API for web
export const loadSumoVersion = () => undefined;
export default {
  crypto_auth,
  crypto_auth_verify,
  crypto_auth_BYTES,
  crypto_auth_KEYBYTES,
  crypto_auth_keygen,
  crypto_aead_xchacha20poly1305_ietf_decrypt,
  crypto_aead_xchacha20poly1305_ietf_encrypt,
  crypto_aead_xchacha20poly1305_ietf_KEYBYTES,
  crypto_aead_xchacha20poly1305_ietf_keygen,
  crypto_aead_xchacha20poly1305_ietf_NPUBBYTES,
  crypto_box_easy,
  crypto_box_seed_keypair,
  crypto_box_keypair,
  crypto_box_NONCEBYTES,
  crypto_box_open_easy,
  crypto_box_PUBLICKEYBYTES,
  crypto_box_SECRETKEYBYTES,
  crypto_generichash,
  crypto_generichash_BYTES,
  crypto_generichash_BYTES_MIN,
  crypto_generichash_BYTES_MAX,
  crypto_generichash_KEYBYTES,
  crypto_generichash_KEYBYTES_MIN,
  crypto_generichash_KEYBYTES_MAX,
  crypto_kdf_derive_from_key,
  crypto_kdf_CONTEXTBYTES,
  crypto_kdf_KEYBYTES,
  crypto_kdf_keygen,
  crypto_pwhash,
  crypto_pwhash_ALG_DEFAULT,
  crypto_pwhash_BYTES_MAX,
  crypto_pwhash_BYTES_MIN,
  crypto_pwhash_MEMLIMIT_INTERACTIVE,
  crypto_pwhash_OPSLIMIT_INTERACTIVE,
  crypto_pwhash_SALTBYTES,
  crypto_secretbox_easy,
  crypto_secretbox_KEYBYTES,
  crypto_secretbox_keygen,
  crypto_secretbox_NONCEBYTES,
  crypto_secretbox_open_easy,
  crypto_sign_detached,
  crypto_sign_keypair,
  crypto_sign_verify_detached,
  from_base64,
  randombytes_buf,
  randombytes_uniform,
  ready,
  to_base64,
  to_hex,
  to_string,
  _unstable_crypto_kdf_hkdf_sha256_BYTES_MAX,
  _unstable_crypto_kdf_hkdf_sha256_BYTES_MIN,
  _unstable_crypto_kdf_hkdf_sha256_KEYBYTES,
  _unstable_crypto_kdf_hkdf_sha256_extract,
  _unstable_crypto_kdf_hkdf_sha256_expand
};
